prometheus.lua

```diff
@@ -320,260 +140,374 @@ local function check_metric_and_label_names(metric_name, label_names)
+  end
+end
+
+-- Construct bucket format for a list of buckets.
 --
--- This should be called once from the `init_by_lua` section in nginx
--- configuration.
+-- This receives a list of buckets and returns a sprintf template that should
+-- be used for bucket boundaries to make them come in increasing order when
+-- sorted alphabetically.
+--
+-- To re-phrase, this is where we detect how many leading and trailing zeros we
+-- need.
 --
 -- Args:
---   dict_name: (string) name of the nginx shared dictionary which will be
---     used to store all metrics
---   prefix: (optional string) if supplied, prefix is added to all
---   metric names on output
+--   buckets: a list of buckets
 --
 -- Returns:
---   an object that should be used to register metrics.
-function Prometheus.init(dict_name, prefix)
-  local self = setmetatable({}, Prometheus)
-  dict_name = dict_name or "prometheus_metrics"
-  self.dict = ngx.shared[dict_name]
-  if self.dict == nil then
-    ngx.log(ngx.ERR,
-      "Dictionary '", dict_name, "' does not seem to exist. ",
-      "Please define the dictionary using `lua_shared_dict`.")
-    return self
-  end
-  self.help = {}
-  if prefix then
-    self.prefix = prefix
-  else
-    self.prefix = ''
+--   (string) a sprintf template.
+local function construct_bucket_format(buckets)
+  local max_order = 1
+  local max_precision = 1
+  for _, bucket in ipairs(buckets) do
+    assert(type(bucket) == "number", "bucket boundaries should be numeric")
+    -- floating point number with all trailing zeros removed
+    local as_string = string.format("%f", bucket):gsub("0*$", "")
+    local dot_idx = as_string:find(".", 1, true)
+    max_order = math.max(max_order, dot_idx - 1)
+    max_precision = math.max(max_precision, as_string:len() - dot_idx)
   end
-  self.type = {}
-  self.registered = {}
-  self.buckets = {}
-  self.bucket_format = {}
-  self.initialized = true
-
-  self:counter("nginx_metric_errors_total",
-    "Number of nginx-lua-prometheus errors")
-  self.dict:set("nginx_metric_errors_total", 0)
-  return self
+  return "%0" .. (max_order + max_precision + 1) .. "." .. max_precision .. "f"
 end
 
-function Prometheus:log_error(...)
-  ngx.log(ngx.ERR, ...)
-  self.dict:incr("nginx_metric_errors_total", 1)
-end
+-- begins metrics functions
+
+local function lookup_or_create(self, label_values)
+  -- if user accidently put a `nil` in between, #label_values will
+  -- return the non-nil prefix of the list, thus we will
+  -- be able to catch that situation as well
+  local cnt = label_values and #label_values or 0
+  -- specially, if first element is nil, # will treat it as "non-empty"
+  if cnt ~= self.label_count or (self.label_count > 0 and not label_values[1]) then
+    return nil, string.format("inconsistent labels count, expected %d, got %d",
+                              self.label_count, cnt)
+  end
+  local t = self.lookup
+  if label_values then
+    -- don't use ipairs here to avoid inner loop generates trace first
```

Comment:
I am not sure I understand what this means. Could you please provide more context?

Reply:
added in comments. for further reading https://konghq.com/blog/building-metrics-pipeline-high-performance-data-collection/ will be a good material.